### PR TITLE
Check the version of the default cgroup mountpoint

### LINKED ISSUE #11094: cgroups version detection fails when systemd.unified_cgroup_hierarchy=1 and default-hierarchy=hybrid
##### Bug report summary

ssia

[cgroups_try_detect_version](https://github.com/netdata/netdata/blob/037584392c9ddb0fecae6ded7c258da783df5aec/collectors/cgroups.plugin/sys_fs_cgroup.c#L138-L198) fails to detect correct Cgroup version and fallbacks to v1.

The problematic part

https://github.com/netdata/netdata/blob/037584392c9ddb0fecae6ded7c258da783df5aec/collectors/cgroups.plugin/sys_fs_cgroup.c#L164-L175

In my case i get `systemd_setting == SYSTEMD_CGROUP_HYBRID`. Seems like our approach is wrong.

---

<details>
<summary>ls -l /sys/fs/cgroup/ [Click]</summary>

```CMD
[ilyam@pc netdata]$ ls -l /sys/fs/cgroup/
total 0
-r--r--r--  1 root root 0 мая  5 13:40 cgroup.controllers
-rw-r--r--  1 root root 0 мая  5 15:00 cgroup.max.depth
-rw-r--r--  1 root root 0 мая  5 15:00 cgroup.max.descendants
-rw-r--r--  1 root root 0 мая  5 13:40 cgroup.procs
-r--r--r--  1 root root 0 мая  5 15:00 cgroup.stat
-rw-r--r--  1 root root 0 мая  5 13:40 cgroup.subtree_control
-rw-r--r--  1 root root 0 мая  5 15:00 cgroup.threads
-rw-r--r--  1 root root 0 мая  5 15:00 cpu.pressure
-r--r--r--  1 root root 0 мая  5 13:40 cpuset.cpus.effective
-r--r--r--  1 root root 0 мая  5 13:40 cpuset.mems.effective
-r--r--r--  1 root root 0 мая  5 15:00 cpu.stat
drwxr-xr-x  2 root root 0 мая  5 14:11 dev-hugepages.mount
drwxr-xr-x  2 root root 0 мая  5 14:11 dev-mqueue.mount
drwxr-xr-x  2 root root 0 мая  5 13:40 init.scope
-rw-r--r--  1 root root 0 мая  5 15:00 io.cost.model
-rw-r--r--  1 root root 0 мая  5 15:00 io.cost.qos
-rw-r--r--  1 root root 0 мая  5 15:00 io.pressure
-r--r--r--  1 root root 0 мая  5 15:00 io.stat
-r--r--r--  1 root root 0 мая  5 15:00 memory.numa_stat
-rw-r--r--  1 root root 0 мая  5 15:00 memory.pressure
-r--r--r--  1 root root 0 мая  5 15:00 memory.stat
drwxr-xr-x  2 root root 0 мая  5 14:11 proc-sys-fs-binfmt_misc.mount
drwxr-xr-x  2 root root 0 мая  5 14:11 sys-fs-fuse-connections.mount
drwxr-xr-x  2 root root 0 мая  5 14:11 sys-kernel-config.mount
drwxr-xr-x  2 root root 0 мая  5 14:11 sys-kernel-debug.mount
drwxr-xr-x  2 root root 0 мая  5 14:11 sys-kernel-tracing.mount
drwxr-xr-x 33 root root 0 мая  5 14:43 system.slice
drwxr-xr-x  3 root root 0 мая  5 14:11 user.slice
```

</details>

<details>
<summary>Relevant logs [Click]</summary>

```CMD
2021-05-05 14:43:49: netdata INFO  : PLUGIN[cgroups] : use unified cgroups false
2021-05-05 14:43:49: netdata ERROR : PLUGIN[cgroups] : CGROUP: cannot find cpuacct mountinfo. Assuming default: /sys/fs/cgroup/cpuacct
2021-05-05 14:43:49: netdata ERROR : PLUGIN[cgroups] : CGROUP: cannot find cpuset mountinfo. Assuming default: /sys/fs/cgroup/cpuset
2021-05-05 14:43:49: netdata ERROR : PLUGIN[cgroups] : CGROUP: cannot find blkio mountinfo. Assuming default: /sys/fs/cgroup/blkio
2021-05-05 14:43:49: netdata ERROR : PLUGIN[cgroups] : CGROUP: cannot find memory mountinfo. Assuming default: /sys/fs/cgroup/memory
2021-05-05 14:43:49: netdata ERROR : PLUGIN[cgroups] : CGROUP: cannot find devices mountinfo. Assuming default: /sys/fs/cgroup/devices
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: cannot read directory '/sys/fs/cgroup/cpuacct' (errno 2, No such file or directory)
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: disabled cpu statistics.
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: cannot read directory '/sys/fs/cgroup/blkio' (errno 2, No such file or directory)
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: disabled blkio statistics.
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: cannot read directory '/sys/fs/cgroup/memory' (errno 2, No such file or directory)
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: disabled memory statistics.
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: cannot read directory '/sys/fs/cgroup/devices' (errno 2, No such file or directory)
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: disabled devices statistics.
```

</details>

##### OS / Environment

```CMD
Linux pc 5.10.32-1-MANJARO #1 SMP Wed Apr 21 14:54:10 UTC 2021 x86_64 GNU/Linux
```

<details>
<summary>release [Click]</summary>

```CMD
/etc/arch-release:Manjaro Linux
/etc/lsb-release:DISTRIB_ID=ManjaroLinux
/etc/lsb-release:DISTRIB_RELEASE=21.0.3
/etc/lsb-release:DISTRIB_CODENAME=Ornara
/etc/lsb-release:DISTRIB_DESCRIPTION="Manjaro Linux"
/etc/manjaro-release:Manjaro Linux
/etc/os-release:NAME="Manjaro Linux"
/etc/os-release:ID=manjaro
/etc/os-release:ID_LIKE=arch
/etc/os-release:BUILD_ID=rolling
/etc/os-release:PRETTY_NAME="Manjaro Linux"
/etc/os-release:ANSI_COLOR="32;1;24;144;200"
/etc/os-release:HOME_URL="https://manjaro.org/"
/etc/os-release:DOCUMENTATION_URL="https://wiki.manjaro.org/"
/etc/os-release:SUPPORT_URL="https://manjaro.org/"
/etc/os-release:BUG_REPORT_URL="https://bugs.manjaro.org/"
/etc/os-release:LOGO=manjarolinux
```

</details>

##### Netdata version

```CMD
Version: netdata v1.30.1-113-g037584392
```

<details>
<summary>buildinfo [Click]</summary>

```CMD
Configure options:  '--prefix=/opt/netdata/usr' '--sysconfdir=/opt/netdata/etc' '--localstatedir=/opt/netdata/var' '--libexecdir=/opt/netdata/usr/libexec' '--libdir=/opt/netdata/usr/lib' '--with-zlib' '--with-math' '--with-user=netdata' '--disable-cloud' 'CFLAGS=-O2' 'LDFLAGS='
Features:
    dbengine:                YES
    Native HTTPS:            YES
    Netdata Cloud:           NO (by user request)
    TLS Host Verification:   YES
Libraries:
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  YES
    libcrypto:               YES
    libm:                    YES
    LWS:                     YES shared-lib
    mosquitto:               YES
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    YES
    cgroup Network Tracking: YES
    CUPS:                    YES
    EBPF:                    YES
    IPMI:                    NO
    NFACCT:                  NO
    perf:                    YES
    slabinfo:                YES
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             NO
    GCP PubSub:              NO
    MongoDB:                 NO
    Prometheus Remote Write: YES
```

</details>

##### Installation method

From source.

##### Component Name

`collectors/cgroups`

##### Steps To Reproduce

- enable Cgroups V2
  - add _systemd.unified_cgroup_hierarchy=1_ to the `GRUB_CMDLINE_LINUX_DEFAULT` (`/etc/default/grub`)
  - update-grub
  - reboot

- start several containers

##### Expected behavior

Cgroups version detection works, it correctly detects Cgroups version when Cgroups V2 is enabled.

### PR DESCRIPTION
**Pull request by vlvkobal**
##### Summary
The PR adds an additional check for the default cgroup mount point. It's a stopgap that fixes incorrect detection of cgroups v1, when `default-hierarchy=hybrid` but only a cgroups v2 hierarchy is mounted.

Fixes #11094

##### Component Name
cgroups plugin

##### Test Plan
Run Netdata with cgroups v2 mounted on `/sys/fs/cgroup` and `default-hierarchy=hybrid`. Unified cgroups should be enabled and working.

### CODE REVIEW

File: collectors/cgroups.plugin/sys_fs_cgroup.c [hunk 1]
```diff
@@ -160,7 +160,18 @@ static enum cgroups_type cgroups_try_detect_version()
     if(!cgroups2_available)
         return CGROUPS_V1;
 
-    // 2. check systemd compiletime setting
+#if defined CGROUP2_SUPER_MAGIC
+    // 2. check filesystem type for the default mountpoint
+    struct statfs fsinfo;
+    if (!statfs("/sys/fs/cgroup", &fsinfo)) {
```
Comment (ilyam8) [inline collectors/cgroups.plugin/sys_fs_cgroup.c:166]: > !statfs("/sys/fs/cgroup")

Should we use `netdata_host_prefix` here?

Comment (vlvkobal) [reply collectors/cgroups.plugin/sys_fs_cgroup.c]: Ups :grin:! Thank you. Fixed :innocent:

**Review by thiagoftsm — APPROVED** [review]

File: collectors/cgroups.plugin/sys_fs_cgroup.c [hunk 2]
```diff
@@ -160,7 +160,20 @@ static enum cgroups_type cgroups_try_detect_version()
     if(!cgroups2_available)
         return CGROUPS_V1;
 
-    // 2. check systemd compiletime setting
+#if defined CGROUP2_SUPER_MAGIC
+    // 2. check filesystem type for the default mountpoint
```
Comment (ilyam8) [inline collectors/cgroups.plugin/sys_fs_cgroup.c:164]: How about making it step 1? Any objections?

Comment (ilyam8) [reply collectors/cgroups.plugin/sys_fs_cgroup.c:164]: ok, not a big deal, nvm then!

Comment (vlvkobal) [reply collectors/cgroups.plugin/sys_fs_cgroup.c:164]: I partly agree with you - there is no need to leave the check of filesystems. But if we want to refine the code, we should delete the whole detection part and make cgroups aware of hybrid configurations.

**Comment by ilyam8** [comment]
Looks good. Thanks for the fix @vlvkobal !

Tested when Netdata installed directly on a host and as a docker container - Cgroup v2 detection works in both cases!

**Review by ilyam8 — APPROVED** [review]

File: collectors/cgroups.plugin/sys_fs_cgroup.c [hunk 3]
```diff
@@ -160,7 +160,20 @@ static enum cgroups_type cgroups_try_detect_version()
...
+    char filename[FILENAME_MAX + 1];
+    snprintfz(filename, FILENAME_MAX, "%s%s", netdata_configured_host_prefix, "/sys/fs/cgroup");
+    struct statfs fsinfo;
+    if (!statfs(filename, &fsinfo)) {
+        if (fsinfo.f_type == CGROUP2_SUPER_MAGIC)
+            return CGROUPS_V2;
+        if (fsinfo.f_type == CGROUP_SUPER_MAGIC)
+            return CGROUPS_V1;
```
Comment (ilyam8) [inline collectors/cgroups.plugin/sys_fs_cgroup.c:172]: Seems like this is never the case. Not a problem. Just want to mention.

```cmd
[ilyam@pc ~]$ mount | grep cgroup
tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,size=4096k,nr_inodes=1024,mode=755,inode64)
cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate)
cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,name=systemd)
cgroup on /sys/fs/cgroup/rdma type cgroup (rw,nosuid,nodev,noexec,relatime,rdma)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
cgroup on /sys/fs/cgroup/net_cls,net_prio type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls,net_prio)
cgroup on /sys/fs/cgroup/perf_event type cgroup (rw,nosuid,nodev,noexec,relatime,perf_event)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
cgroup on /sys/fs/cgroup/hugetlb type cgroup (rw,nosuid,nodev,noexec,relatime,hugetlb)
cgroup on /sys/fs/cgroup/pids type cgroup (rw,nosuid,nodev,noexec,relatime,pids)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
cgroup on /sys/fs/cgroup/memory type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
cgroup on /sys/fs/cgroup/cpu,cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpu,cpuacct)
```

**Pull request merged**